api/include/opentelemetry/logs/log_record.h

```diff
@@ -82,25 +77,28 @@ struct LogRecord
   core::SystemTimestamp timestamp;  // uint64 nanoseconds since Unix epoch
   trace::TraceId trace_id;          // byte sequence
   trace::SpanId span_id;            // byte sequence
-  trace::TraceFlags trace_flag;     // byte
-  Severity severity;  // Severity enum that combines severity_text and severity_number in the
-                      // LogDataModel (can separate in SDK)
+  trace::TraceFlags trace_flags;    // byte
+  Severity severity;                // Severity enum that combines severity_text and severity_number
 
   // other fields that will not be set by default
   nostd::string_view name;  // string
   nostd::string_view body;  // currently a simple string, but should be changed "Any" type
-  common::KeyValueIterable &resource;    // key/value pair list
-  common::KeyValueIterable &attributes;  // key/value pair list
+  nostd::shared_ptr<common::KeyValueIterable> resource;    // key/value pair list
+  nostd::shared_ptr<common::KeyValueIterable> attributes;  // key/value pair list
```

Comment:
Can't we avoid this shared ownership by overloaded Log constructor to accept reference to kv-iterable. 

btw, not related to this PR but this is where we would have benefited from `Recordable` interface - Instead of storing these attributes/resources here, we could have directly propagated them to exporters.

Reply:
I was thinking that since `LogRecord` is what is storing the information of the log from the API all the way to the Exporters, then having the log record have shared ownership over the resource and attributes would make sense here. If it's just a reference, and the initial attributes/resources goes out of scope and gets deallocated while the record is still in the processor/exporter, then that information would be lost which would be bad. Yea when the `Recordable` interface is implemented it would definitely simplify things.

Reply:
One reason for the recordable interface is to ensure the Exporter can allocate the record.  This allows, e.g. an Exporter to pre-allocate a chunk of memory and ensure *all* logs fit within that limited amount of memory rather than contesting with the HEAP of the main application.

Two questions:

1. Should "resource" be mutable at all, or can we ensure there is one when creating the LogRecord?
  - If `LogRecord` has a resource attached later in the API than when it's created, I understand it would have to be a some kind of pointer.
  - Are Log API users able to modify the "Resource" object itself?  What's the use case for that?
2. Shouldn't the LogRecord own the "attributes"? (i.e. not be a shared pointer at all)?   When do you expect these to be shared?

Reply:
For the resources and attributes, it probably doesn't need to be mutable. When creating a LogRecord, I made it a shared_ptr such that the user wouldn't need to pass it into the constructor, but could set it afterwards like the other LogRecord fields (for consistency). 

And for your second question, the LogRecord should definitely own the attributes. Instead of changing the LogRecord/making a recordable interface, for now I could simply duplicate the memory from the attributes into the SDK heap, and the LogRecord's shared_ptr would point to that chunk. Since there would be no other shared_ptr to that object, it would technically "own" it as it would get deallocated when the LogRecord gets destructed. This isn't the most optimal approach, and in the future the LogRecord could directly contain the attributes and resources, but I think using shared_ptr would work to get the basic prototype out and running.

Reply:
> but I think using shared_ptr would work to get the basic prototype out and running.

I think it would be important to do it right at first go. Based on the answers of the questions put up by @jsuereth , it doesn't look that the shared_ptr is the right approach of doing this. We should either 
  - transfer the ownership of attributes/resources to LogRecord.
  -  Keep using references to these attributes/reference. What challenge you see here ?

Reply:
Yea sure, I'll convert it back to using references since its less overhead than using shared/unique ptr.

Reply:
One issue I'm thinking of is if we do it this way without transferring the ownership, then the LogRecord is going to be dependant on the attributes/resource object within the user's code. If the user's object gets changed or deleted while the LogRecord is inside the processor or exporter, then this information is going to be lost.

We can solve this by doing a memcpy in the LogRecord so that it owns the memory of the attributes/resources, but it would also decrease performance... I'll leave it as reference for now but if anyone has any strong opinions about this I can take a look at changing it.

Reply:
> If the user's object gets changed or deleted while the LogRecord is inside the processor or exporter, then this information is going to be lost.

Correct, that can happen if User creates `LogRecord` object instance, deletes the attribute (or it goes out of scope), and then calls actual logger method.

Reply:
And if that is the case, would prefer to have the overhead of unique/shared ptr , than to see the actual application crashing. Would wait for others to comment cc @jsuereth @ThomsonTan @pyohannes